includes/sanitizers/class-amp-layout-sanitizer.php

```diff
@@ -0,0 +1,51 @@
+<?php
+/**
+ * Class AMP_Layout_Sanitizer
+ *
+ * @since 1.5.0
+ * @package AMP
+ */
+
+/**
+ * Class AMP_Layout_Sanitizer
+ *
+ * @since 1.5.0
+ */
+class AMP_Layout_Sanitizer extends AMP_Base_Sanitizer {
+
+	/**
+	 * Sanitize any element that has the `layout` or `data-amp-layout` attribute.
+	 */
+	public function sanitize() {
+		$xpath = new DOMXPath( $this->dom );
+		$nodes = $xpath->query( '//*[ @layout or @data-amp-layout ]' );
+
+		foreach ( $nodes as $node ) {
+			$width  = $node->getAttribute( 'width' );
+			$height = $node->getAttribute( 'height' );
+
+			// The `layout` attribute can also be defined through the `data-amp-layout` attribute.
+			if ( $node->hasAttribute( 'data-amp-layout' ) ) {
+				$layout = $node->getAttribute( 'data-amp-layout' );
+				$node->setAttribute( 'layout', $layout );
+				$node->removeAttribute( 'data-amp-layout' );
+			}
+
+			// If the width & height are `100%` the layout must be `fill`.
+			if ( '100%' === $width && '100%' === $height ) {
+				$node->removeAttribute( 'width' );
+				$node->removeAttribute( 'height' );
+				$node->setAttribute( 'layout', AMP_Rule_Spec::LAYOUT_FILL );
+				return;
+			}
+
+			// If the width is `100%`, convert the layout to `fixed-height` and width to `auto`.
+			if ( '100%' === $width ) {
+				$node->setAttribute( 'width', 'auto' );
+				$node->setAttribute( 'layout', AMP_Rule_Spec::LAYOUT_FIXED_HEIGHT );
+			}
```

Comment:
Inferring the `layout=fill` and `layout=fixed-height` need to be done on elements that _don't_ already have the `layout` defined.

In particular, these are needed for converting non-AMP elements (which don't have `layout`) into AMP elements that _do_ have `layout`.

Reply:
In other words, if someone has provided an invalid `layout` attribute for a given width/height, then then this should freely result in an `invalid_layout` validation error. If someone is writing AMP then they need to be told they are doing it wrong.

Reply:
Done in https://github.com/ampproject/amp-wp/pull/3728/commits/18c0833f63a9eebf054ba067e67bef3a585a6252.

Reply:
This would then mean that the fix for #2146 is to remove the embed, as the layout would be invalid.

Reply:
For any element that that has `width=100%` we should always convert this into a `fixed-height` layout. I see that the Facebook comment is a special case because it has a `data-width` attribute, not just a `width`. So yes, that conversion would have to go into the Facebook sanitizer. 

But otherwise, if encountering any element with `width=100%` it should be converted into `width=auto layout=fixed-height`. 

But, if the element already has a `layout` attribute specified, then this indicates the author says they are (trying) to write AMP, and in that case, we should go ahead and raise an `INVALID_LAYOUT` error and remove the element. This is because we don't know how to recover from the error (as we could override the `layout` the user supplied), and the author wouldn't be informed of their error. Since the site is attempting to write AMP markup, then we should apply the validation rules and tell them they are doing it wrong.

Reply:
Thanks for the detailed explanation. A simple fix for that Facebook comment could be to simply use `data-layout` when setting the node's attributes, then let the layout sanitizer take care of it:

```diff
diff --git a/includes/embeds/class-amp-facebook-embed.php b/includes/embeds/class-amp-facebook-embed.php
index d10ec8dc..8f5ff75a 100644
--- a/includes/embeds/class-amp-facebook-embed.php
+++ b/includes/embeds/class-amp-facebook-embed.php
@@ -195,7 +195,7 @@ class AMP_Facebook_Embed_Handler extends AMP_Base_Embed_Handler {
        private function create_amp_facebook_and_replace_node( $dom, $node, $embed_type ) {
 
                $attributes = [
-                       'layout' => 'responsive',
+                       'data-layout' => 'responsive',
                        'width'  => $node->hasAttribute( 'data-width' ) ? $node->getAttribute( 'data-width' ) : $this->DEFAULT_WIDTH,
                        'height' => $node->hasAttribute( 'data-height' ) ? $node->getAttribute( 'data-height' ) : $this->DEFAULT_HEIGHT,
                ];
```